homeassistant/components/signal_messenger/notify.py

```diff
@@ -43,37 +55,93 @@ def get_service(hass, config, discovery_info=None):
 class SignalNotificationService(BaseNotificationService):
     """Implement the notification service for SignalMessenger."""
 
-    def __init__(self, recp_nrs, signal_cli_rest_api):
+    def __init__(
+        self,
+        recp_nrs: list[str],
+        signal_cli_rest_api: SignalCliRestApi,
+    ) -> None:
         """Initialize the service."""
 
         self._recp_nrs = recp_nrs
         self._signal_cli_rest_api = signal_cli_rest_api
 
-    def send_message(self, message="", **kwargs):
-        """Send a message to a one or more recipients.
-
-        Additionally a file can be attached.
-        """
+    def send_message(self, message: str = "", **kwargs: Any) -> None:
+        """Send a message to a one or more recipients. Additionally a file can be attached."""
 
         _LOGGER.debug("Sending signal message")
 
         data = kwargs.get(ATTR_DATA)
 
-        filenames = None
-        if data is not None:
-            if ATTR_FILENAMES in data:
-                filenames = data[ATTR_FILENAMES]
-            if ATTR_FILENAME in data:
-                _LOGGER.warning(
-                    "The 'attachment' option is deprecated, please replace it with 'attachments'. This option will become invalid in version 0.108"
-                )
-                if filenames is None:
-                    filenames = [data[ATTR_FILENAME]]
-                else:
-                    filenames.append(data[ATTR_FILENAME])
+        filenames = self.get_filenames(data)
+        attachments_as_bytes = self.get_attachments_as_bytes(
+            data, CONF_MAX_ALLOWED_DOWNLOAD_SIZE_BYTES
+        )
 
         try:
-            self._signal_cli_rest_api.send_message(message, self._recp_nrs, filenames)
+            self._signal_cli_rest_api.send_message(
+                message, self._recp_nrs, filenames, attachments_as_bytes
+            )
         except SignalCliRestApiError as ex:
             _LOGGER.error("%s", ex)
             raise ex
+
+    @staticmethod
+    def get_filenames(data: Any) -> list[str] | None:
+        """Extract attachment filenames from data."""
+        if data is None:
+            return None
+
+        if ATTR_FILENAMES in data:
+            if isinstance(data[ATTR_FILENAMES], list):
+                return data[ATTR_FILENAMES]
+
+            raise ValueError(f"'{ATTR_FILENAMES}' property must be a list")
+
+        return None
+
+    @staticmethod
+    def get_attachments_as_bytes(
+        data: Any, attachment_size_limit: int
+    ) -> list[bytearray] | None:
```

Comment:
IMHO this is a no-go.
It downloads files from a possible random URL and sends it out.

Reply:
> IMHO this is a no-go. It downloads files from a possible random URL and sends it out as attachment.

What are your specific concerns with this approach? If we were to only support images, videos and audio would that be acceptable?

Reply:
My concern is that using a call a random file can be downloaded and shipped off our platform. That sounds not acceptable to me.

Reply:
The HA Companion app shows an example of loading an attachment from Github - is this different in some way?

https://companion.home-assistant.io/docs/notifications/notification-attachments#example-service-call

Reply:
Yes, as those are not sent over as binary data.
In that case, the app receiving is responsible for handling external data responsibly.

Reply:
I would love any advice you or anyone else may have on how we can achieve our goals of attaching files from an API to the message in a safe and acceptable way. For this use case I am looking to attach images and videos from the [Frigate Notification API](https://docs.frigate.video/integrations/home-assistant#notification-api) and sending them to a Signal message.

It appears that the Telegram integration does this the same way as we have designed:
https://www.home-assistant.io/integrations/telegram/#photo-support
https://github.com/home-assistant/core/blob/dev/homeassistant/components/telegram_bot/__init__.py#L257

Reply:
I agree it does seem like a security risk, but in reality, home assistant is a hub and is going to connect to external sources. Sure, the code to download the data could be moved to the library (as in the mobile app), but it's still home assistant that is running the code from the library. 

An alternative is to limit this to the home assistant external/internal URL's or provide a whitelist of sorts, however that would greatly limit functionality. Not to mention that as @burnsie-la mentioned the Telegram integration already does this, and I would not be surprised if other components do this as well. Perhaps there would need to be a greater architecture change around home assistant as a whole to address @frenck's comment.

Reply:
@burnsie-la It seems like the best way to protect the user and the platform would be to implement it like the Slack integration with a URL whitelist defined in the home assistant core config. 

See below:
https://www.home-assistant.io/integrations/slack/

https://www.home-assistant.io/docs/configuration/basic/#allowlist_external_urls

Reply:
@burnsie-la The above solution/suggestion by @alexyao2015 is indeed the right solution to this, which can/would be accepted.

Reply:
Thanks - now added in 9d6d7b1